### PR TITLE
ogre 1.10.9 (New Formula)

### PR DESCRIPTION
**Pull request by s172262**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

### CODE REVIEW

**Comment by s172262** [comment]
@BrewTestBot test this please

File: Formula/ogre.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    inreplace "CMake/Utils/OgreConfigTargets.cmake", "set(PLATFORM_NAME \"macosx\")", ""
+    inreplace "CMake/CMakeLists.txt", "set(OGRE_CMAKE_DIR \"CMake\")", "set(OGRE_CMAKE_DIR \"lib/OGRE/cmake\")"
+    inreplace ["Samples/Media/CMakeLists.txt", "CMake/InstallResources.cmake"],
+              "set(OGRE_MEDIA_PATH \"Media\")", "set(OGRE_MEDIA_PATH \"share/ogre/Media\")"
+    inreplace "CMake/InstallResources.cmake", "set(OGRE_CFG_INSTALL_PATH \"bin\")", "set(OGRE_CFG_INSTALL_PATH \"share/ogre/cfg\")"
+
+    args = std_cmake_args
+    args += %W[-DCMAKE_SKIP_BUILD_RPATH=FALSE
+               -DOGRE_FULL_RPATH=TRUE
+               -DOGRE_LIB_DIRECTORY=#{lib}
```
Comment (scpeters) [inline Formula/ogre.rb]: I think `#{lib}` is an absolute path, but there are places in the ogre code that expect `OGRE_LIB_DIRECTORY` to not be absolute, for example the pkgconfig templates:

* https://github.com/OGRECave/ogre/blob/v1.10.9/CMake/Templates/OGRE.pc.in#L3
* https://github.com/OGRECave/ogre/blob/v1.10.9/CMake/Templates/OGRE-RTShaderSystem.pc.in#L3
* etc.

This causes the following warnings when compiling against ogre:

~~~
ld: warning: directory not found for option
 '-L/usr/local/Cellar/ogre/1.10.9/usr/local/Cellar/ogre/1.10.9/lib'
~~~

Comment (s172262) [reply Formula/ogre.rb]: Thank you for your advice.

Comment (s172262) [reply Formula/ogre.rb]: It's O.K.
```
$ brew list ogre
/usr/local/Cellar/ogre/1.10.9/bin/OgreMeshUpgrader
/usr/local/Cellar/ogre/1.10.9/bin/OgreXMLConverter
/usr/local/Cellar/ogre/1.10.9/bin/VRMLConverter
/usr/local/Cellar/ogre/1.10.9/include/OGRE/ (574 files)
/usr/local/Cellar/ogre/1.10.9/lib/libOgreBites.1.10.9.dylib
/usr/local/Cellar/ogre/1.10.9/lib/libOgreHLMS.1.10.9.dylib
/usr/local/Cellar/ogre/1.10.9/lib/libOgreMain.1.10.9.dylib
/usr/local/Cellar/ogre/1.10.9/lib/libOgreMeshLodGenerator.1.10.9.dylib
/usr/local/Cellar/ogre/1.10.9/lib/libOgreOverlay.1.10.9.dylib
/usr/local/Cellar/ogre/1.10.9/lib/libOgrePaging.1.10.9.dylib
/usr/local/Cellar/ogre/1.10.9/lib/libOgreProperty.1.10.9.dylib
/usr/local/Cellar/ogre/1.10.9/lib/libOgreRTShaderSystem.1.10.9.dylib
/usr/local/Cellar/ogre/1.10.9/lib/libOgreTerrain.1.10.9.dylib
/usr/local/Cellar/ogre/1.10.9/lib/libOgreVolume.1.10.9.dylib
/usr/local/Cellar/ogre/1.10.9/lib/OGRE/ (23 files)
/usr/local/Cellar/ogre/1.10.9/lib/pkgconfig/ (11 files)
/usr/local/Cellar/ogre/1.10.9/lib/python2.7/ (9 files)
/usr/local/Cellar/ogre/1.10.9/lib/ (11 other files)
/usr/local/Cellar/ogre/1.10.9/share/ogre/ (944 files)
```

Comment (scpeters) [reply Formula/ogre.rb]: I meant that the configured pkg-config files are incorrect:

~~~
$ grep 'libdir=' /usr/local/Cellar/ogre/1.10.9/lib/pkgconfig/OGRE.pc 
libdir=${prefix}//usr/local/Cellar/ogre/1.10.9/lib
~~~

File: Formula/ogre.rb [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      You might add "#{HOMEBREW_PREFIX}/lib/python2.7/dist-packages" to PYTHONPATH as following.
+
+        export PYTHONPATH="#{HOMEBREW_PREFIX}/lib/python2.7/dist-packages:$PYTHONPATH"
+
+      EOS
+    end
+  end
+
+  test do
+    system "#{bin}/OgreXMLConverter", "-v"
```
Comment (woodruffw) [inline Formula/ogre.rb]: New homebrew-core formulae are required to have a more substantive test than just `version` or `help`.

Could you come up with a minimal functional test for this package?

File: Formula/ogre.rb [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+    system "cmake", ".", *args
+    system "make", "install"
+    if build.with? "java-component"
+      lib.install "java/libs/libOgreJNI.jnilib"
+    end
+  end
+
+  def caveats
+    if build.with? "python-component" then <<~EOS
```
Comment (woodruffw) [inline Formula/ogre.rb]: This is generic advice and not specific to this formula, so you can drop this `caveats`.

File: Formula/ogre.rb (hunk 2 above)
Comment (scpeters) [reply Formula/ogre.rb]: We have a more substantive test in our tap for an older version of ogre:

* https://github.com/osrf/homebrew-simulation/blob/master/ogre1.9.rb#L113-L143

File: Formula/ogre.rb (hunk 3 above)
Comment (s172262) [reply Formula/ogre.rb]: I feel this is rare case. 
```
$ find /usr/local/Cellar/ -name "dist-packages"
/usr/local/Cellar//ogre/1.10.9/lib/python2.7/dist-packages
$
```

File: Formula/ogre.rb (hunk 2 above)
Comment (s172262) [reply Formula/ogre.rb]: Thank you! I would use it.

File: Formula/ogre.rb (hunk 1 above)
Comment (s172262) [reply Formula/ogre.rb]: I am sorry. But `OGRE.pc` is correct now.
```
$ cat /usr/local/Cellar/ogre/1.10.9/lib/pkgconfig/OGRE.pc 
prefix=/usr/local/Cellar/ogre/1.10.9
exec_prefix=${prefix}
libdir=${prefix}/lib
includedir=${prefix}/include
plugindir=${libdir}/OGRE

Name: OGRE
Description: Object-Oriented Graphics Rendering Engine
Version: 1.10.9
URL: http://www.ogre3d.org
Libs: -L${libdir} -lOgreMain  -lpthread
Cflags: -I${includedir} -I${includedir}/OGRE -I${includedir}/OGRE/OSX -pthread
```

File: Formula/ogre.rb [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+  if build.with?("java-component") || build.with?("python-component")
+    depends_on "swig"
+    depends_on :java if build.with? "java-component"
+    depends_on :python if build.with? "python-component"
+  end
+
+  def install
+    if build.with? "boost"
+      unless Tab.for_name("boost").cxx11?
+        odie "Boost build with \"--c++11\" is needed because this formula needs c++11."
```
Comment (dunn) [inline Formula/ogre.rb]: I believe correct way to do this is
```ruby
depends_on "boost" => [:optional, :cxx11]
```

Comment (ilovezfs) [reply Formula/ogre.rb]: This isn't permitted at all actually. It needs to use the standard boost.

File: Formula/ogre.rb [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  sha256 "85ba2cc6a35c67ff93a9a498af9f8f3113fd3a16e7cc43c18b9769d8bf1e9101"
+
+  option "with-java-component", "Build with java component"
+  option "with-python-component", "Build with python component"
+
+  needs :cxx11
+
+  depends_on "cmake" => :build
+  depends_on "pkg-config" => :build
+  depends_on "boost" => [:optional, :cxx11]
```
Comment (ilovezfs) [inline Formula/ogre.rb]: Needs to be removed.

File: Formula/ogre.rb [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    ENV.cxx11
+    ENV.append "CXXFLAGS", "-std=c++11"
+    if ENV.compiler == :clang
+      ENV.append "CXXFLAGS", "-stdlib=libc++"
+      ENV.append "LINKFLAGS", "-stdlib=libc++"
+    end
+    # OGRE tends to use "/Library/Frameworks/SDL.framework" if exist. Because OGRE appends "-F/Library/Frameworks".
+    ENV.prepend "CPPFLAGS", "-I#{Formula["sdl"].opt_include}/SDL"
+
+    inreplace "CMake/Utils/OgreConfigTargets.cmake", "set(PLATFORM_NAME \"macosx\")", ""
```
Comment (ilovezfs) [inline Formula/ogre.rb]: string replacements like this aren't allowed in new formulae.

File: Formula/ogre.rb [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    # Trunk starts using "clang++ -x c" to select C compiler which breaks C++11
+    # handling using ENV.cxx11. Using "cxxflags" and "linkflags" still works.
+    ENV.cxx11
+    ENV.append "CXXFLAGS", "-std=c++11"
+    if ENV.compiler == :clang
+      ENV.append "CXXFLAGS", "-stdlib=libc++"
+      ENV.append "LINKFLAGS", "-stdlib=libc++"
+    end
+    # OGRE tends to use "/Library/Frameworks/SDL.framework" if exist. Because OGRE appends "-F/Library/Frameworks".
+    ENV.prepend "CPPFLAGS", "-I#{Formula["sdl"].opt_include}/SDL"
```
Comment (ilovezfs) [inline Formula/ogre.rb]: This is a bug that needs to be fixed upstream

File: Formula/ogre.rb [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+      inreplace "Components/Java/CMakeLists.txt", "set(CMAKE_SWIG_FLAGS -w401,314 -package org.Ogre)",
+                "set(CMAKE_SWIG_FLAGS -w401,314 -package org.Ogre -D_LIBCPP_VERSION)"
+    else
+      args << "-DOGRE_BUILD_COMPONENT_JAVA=OFF"
+    end
+
+    system "cmake", ".", *args
+    system "make", "install"
+    lib.install "java/libs/libOgreJNI.jnilib" if build.with? "java-component"
+    MachO::Tools.change_install_name("#{lib}/OGRE/Plugin_OctreeZone.#{version}.dylib",
```
Comment (ilovezfs) [inline Formula/ogre.rb]: This kind of manipulation is not allowed in a new formula.

File: Formula/ogre.rb [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    ENV.append "CXXFLAGS", "-std=c++11"
+    if ENV.compiler == :clang
+      ENV.append "CXXFLAGS", "-stdlib=libc++"
+      ENV.append "LINKFLAGS", "-stdlib=libc++"
+    end
+    # OGRE tends to use "/Library/Frameworks/SDL.framework" if exist. Because OGRE appends "-F/Library/Frameworks".
+    ENV.prepend "CPPFLAGS", "-I#{Formula["sdl"].opt_include}/SDL"
+
+    inreplace "CMake/Utils/OgreConfigTargets.cmake", "set(PLATFORM_NAME \"macosx\")", ""
+    inreplace "CMake/CMakeLists.txt", "set(OGRE_CMAKE_DIR \"CMake\")", "set(OGRE_CMAKE_DIR \"lib/OGRE/cmake\")"
```
Comment (ilovezfs) [inline Formula/ogre.rb]: ditto.

File: Formula/ogre.rb [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    inreplace ["Samples/Media/CMakeLists.txt", "CMake/InstallResources.cmake"],
```
Comment (ilovezfs) [inline Formula/ogre.rb]: ditto.

File: Formula/ogre.rb [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+              "set(OGRE_MEDIA_PATH \"Media\")", "set(OGRE_MEDIA_PATH \"share/ogre/Media\")"
+    inreplace "CMake/InstallResources.cmake", "set(OGRE_CFG_INSTALL_PATH \"bin\")", "set(OGRE_CFG_INSTALL_PATH \"share/ogre/cfg\")"
```
Comment (ilovezfs) [inline Formula/ogre.rb]: ditto.

File: Formula/ogre.rb [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+               -DOGRE_BUILD_RENDERSYSTEM_GL3PLUS=OFF
+               -DOGRE_BUILD_SAMPLES=OFF
+               -DOGRE_BUILD_PLUGIN_CG=OFF
+               -DOGRE_BUILD_LIBS_AS_FRAMEWORKS=OFF]
+    # "std" meands using c++11.
+    args << "-DOGRE_CONFIG_THREAD_PROVIDER=" + (build.with?("boost") ? "boost" : "std")
+
+    # See https://github.com/OGRECave/ogre/issues/563
+    if build.with?("python-component")
+      inreplace "Components/Python/CMakeLists.txt", "set(CMAKE_SWIG_FLAGS -w401,314 -builtin)",
```
Comment (ilovezfs) [inline Formula/ogre.rb]: ditto.

File: Formula/ogre.rb [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+                "set(CMAKE_SWIG_FLAGS -w401,314 -builtin -D_LIBCPP_VERSION)"
+    else
+      args << "-DOGRE_BUILD_COMPONENT_PYTHON=OFF"
+    end
+    if build.with?("java-component")
+      inreplace "Components/Java/CMakeLists.txt", "set(CMAKE_SWIG_FLAGS -w401,314 -package org.Ogre)",
```
Comment (ilovezfs) [inline Formula/ogre.rb]: ditto.

**Comment by s172262** [comment]
O.K. I will close.

**Pull request closed without merging**